packages/common/src/utils/abort-signal-utils.ts

```diff
@@ -0,0 +1,38 @@
+import { timer, fromEvent, merge } from 'rxjs'
+import { first } from 'rxjs/operators'
+
+export function mergeAbortSignals(signals: AbortSignal[]): AbortSignal {
+  const controller = new AbortController()
+
+  if (signals.length === 0) {
+    throw Error('Need abort signals to create a merged abort signal')
+  }
+
+  if (signals.some((signal) => signal.aborted)) {
+    controller.abort()
+    return controller.signal
+  }
+
+  merge(...signals.map((signal) => fromEvent(signal, 'abort')))
+    .pipe(first())
+    .subscribe(() => {
+      controller.abort()
+    })
+
+  return controller.signal
+}
+
+export interface TimedAbortSignal {
+  signal: AbortSignal
+  clear: () => void
+}
+
+export function createTimedAbortSignal(timeout: number): TimedAbortSignal {
```

Comment:
Weak suggestion: use class instead of function and interface. I do not know why, just have a feeling, it might be easier to support.